common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java

```diff
@@ -0,0 +1,91 @@
+
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.adal.internal.net;
+
+import android.content.Context;
+import android.net.ConnectivityManager;
+import android.os.Build;
+
+import androidx.test.core.app.ApplicationProvider;
+
+import com.microsoft.identity.common.adal.internal.PowerManagerWrapperShadow;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+import org.robolectric.util.ReflectionHelpers;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(shadows = {
+        PowerManagerWrapperShadow.class
+})
+public class DefaultConnectionServiceTest {
+
+    protected Context mContext;
+
+    @Before
+    public void setup() {
+        mContext = ApplicationProvider.getApplicationContext();
+    }
+
+    @Test
+    public void testConnectionAvailableWhenConnected() {
+        final DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
+
+        assertTrue(defaultConnectionService.isConnectionAvailable());
+    }
+
+    @Test
+    public void testConnectionNotAvailableWhenNoActiveNetwork() {
+        final ConnectivityManager connectivityManager = (ConnectivityManager) mContext
+                .getSystemService(Context.CONNECTIVITY_SERVICE);
+
+        shadowOf(connectivityManager).setActiveNetworkInfo(null);
+        final DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
+
+        assertFalse(defaultConnectionService.isConnectionAvailable());
+    }
```

Comment:
What I would suggest here is to trigger a few different network events, if possible.  If a single network is unavailable, we should still be able to connect, or if we lose one connection and gain another we should remain connected.

Reply:
I think the following permutations would be sufficient:

Have 2, lose 1, stay connected.
Have 2, lose both, not connected, gain 1, stay connected.
Have none, not connected, gain 1, are connected.
Have none, gain 2, are connected, lose 1, are connected.

The idea being that we can then bring back the changes to be event driven and have these tests tell us that they're safe.

Reply:
Brilliant!! ... Thanks Adam.